core/src/main/java/org/apache/iceberg/hadoop/SerializedConfiguration.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.hadoop;
+
+import java.io.Serializable;
+import java.util.Map;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+/**
+ * A class that captures the current state of a Hadoop configuration in a serializable manner.
+ * <p>
+ * <em>Note:</em> Unlike {@link SerializableConfiguration}, no subsequent changes to the underlying
+ * Hadoop configuration will be propagated to the captured state once an instance
+ * of {@link SerializedConfiguration} is constructed.
+ */
+public class SerializedConfiguration implements Serializable {
```

Comment:
We cannot just replace the internal implementation of the existing `SerializableConfiguration` class as Hadoop conf is not immutable and can dynamically change. For example, getting `FileSystem` would trigger a full reload of configs. That's why I added `SerializedConfiguration` that is used only in `SerializedTable` for now. In the future, we can use `SerializedConfiguration` in `writeReplace` of `SerializableConfiguration`.

Reply:
This way reduced the size of the serialized Hadoop conf by 50% as we don't serialize the source of the conf.

Reply:
Also, this class is Kryo compatible. So we don't have to wrap it into special classes like we do today in Spark.